### PR TITLE
Add `-mount` flag to kv list command

### PR DESCRIPTION
**Pull request by dhuckins**
resolves #17399

### CODE REVIEW

File: command/kv_list.go [hunk 1]
```diff
@@ -75,25 +93,56 @@ func (c *KVListCommand) Run(args []string) int {
...
 	}
 
-	// Sanitize path
-	path := sanitizePath(args[0])
-	mountPath, v2, err := isKVv2(path, client)
-	if err != nil {
-		c.UI.Error(err.Error())
-		return 2
-	}
+	// If true, we're working with "-mount=secret foo" syntax.
```
Comment (dhuckins) [inline command/kv_list.go:96]: logic is the same as in kv put
https://github.com/hashicorp/vault/blob/d35be2d0de3d1c036248570c538c2051c4c1dc57/command/kv_put.go#L144C2-L176

File: command/kv_test.go [hunk 2]
```diff
@@ -590,6 +590,97 @@ func TestKVGetCommand(t *testing.T) {
...
+
+// TestKVListCommand runs tests for `vault kv list`
+func TestKVListCommand(t *testing.T) {
+	testCases := []struct {
+		name       string
+		args       []string
+		outStrings []string
+		code       int
+	}{
+		{
```
Comment (lursu) [inline command/kv_test.go:613]: Is it worth adding a V1 test cases since we technically have a logical chage for V1 vs V2?

**Review by lursu — APPROVED** [review]
One question about test coverage but otherwise LGTM.

**Review by AnPucel — APPROVED** [review]

File: command/kv_test.go (hunk 2 above)
Comment (dhuckins) [reply command/kv_test.go:613]: good call! let me see how the other tests did it

File: command/kv_list.go [hunk 3]
```diff
@@ -40,7 +42,23 @@ Usage: vault kv list [options] PATH
...
+
+	f.StringVar(&StringVar{
+		Name:    "mount",
+		Target:  &c.flagMount,
+		Default: "", // no default, because the handling of the next arg is determined by whether this flag has a value
+		Usage: `Specifies the path where the KV backend is mounted. If specified, 
+		the next argument will be interpreted as the secret path. If this flag is 
+		not specified, the next argument will be interpreted as the combined mount 
+		path and secret path, with /data/ automatically appended between KV 
+		v2 secrets.`,
```
Comment (averche) [inline command/kv_list.go:58]:
```suggestion
		path and secret path, with /data/ automatically inserted between the two
		paths for KV-v2 secrets.`,
```

File: command/kv_list.go [hunk 4]
```diff
@@ -75,25 +93,56 @@ func (c *KVListCommand) Run(args []string) int {
...
-	secret, err := client.Logical().List(path)
+	// Add /data to v2 paths only
+	var fullPath string
+	if v2 {
+		fullPath = addPrefixToKVPath(partialPath, mountPath, "metadata")
+	} else {
+		// v1
+		if mountFlagSyntax {
+			fullPath = path.Join(mountPath, partialPath)
+		} else {
```
Comment (averche) [inline command/kv_list.go:138]: It's a bit hard to follow these if blocks, perhaps they can be rearranged a bit for clarity? In particular we are doing `path.Join(...)` for kv-v1 here yet for kv-v2 in the previous `if` block.

**Review by averche — APPROVED** [review]
A couple small nits, otherwise looks good.

File: command/kv_list.go (hunk 3 above)
Comment (dhuckins) [reply command/kv_list.go:58]: this is copied from `kv get` 
https://github.com/hashicorp/vault/blob/e4e9612fa42edc5e66c5c1528c65334da8c76062/command/kv_get.go#L72-L76
so going to leave for now

Comment (averche) [reply command/kv_list.go:58]: Perhaps adjusting both descriptions for clarity would help.

File: command/kv_list.go (hunk 4 above)
Comment (dhuckins) [reply command/kv_list.go:138]: just copied the logic from [kv put](https://github.com/hashicorp/vault/blob/e4e9612fa42edc5e66c5c1528c65334da8c76062/command/kv_put.go#L154-L198)

each command in `kv ...` does this a little differently, going to look to consolidate the logic in a subsequent ticket

**Review by digivava — CHANGES_REQUESTED** [review]
Thanks for taking a look at this! The reason that I didn't have a `-mount` flag for this command is not that I forgot it so much as it behaved a little differently from the others so I just kinda skipped it, but I think it's tripped up a few people so it's definitely worth adding.

Here's the thing about this command: the `kv list` command is the only one that lets you pass just the mount path itself with no secret path at all. When you do this, it shows all the secrets at that mount path, and indeed I think this is a common use case for the command.

For example, let's say I have some secrets I've created inside the `secret` mount path. 
If I run your code with the traditional way of passing the path:

```
$ vault kv list secret
Keys
----
foo
bar
```

```
$ vault kv list -mount=secret
Not enough arguments (expected 1, got 0)
```

So basically, if you can fix it so that version of the command doesn't throw an error, and instead returns the same thing as the version without the flag, then I think this will be good to go!

**Comment by averche** [comment]
> ```
> $ vault kv list secret
> Keys
> ----
> foo
> bar
> ```
> 
> vs with the -mount flag:
> 
> ```
> $ vault kv list -mount=secret
> Not enough arguments (expected 1, got 0)
> ```

Would it be reasonable to require a "/" or a "." parameter to be provided for this edge case:

```sh
vault kv list -mount=secret /
```

File: command/kv_test.go [hunk 5]
```diff
@@ -590,6 +590,113 @@ func TestKVGetCommand(t *testing.T) {
...
+			// this is behavior that should be tested
+			// `kv` here is an explicit mount
+			// `my-prefix` is not
+			// the current kv code will ignore `my-prefix`
+			name:       "ignore_multi_part_mounts",
+			args:       []string{"-mount", "kv/my-prefix"},
+			outStrings: []string{"my-prefix"},
+			code:       0,
+		},
```
Comment (dhuckins) [inline command/kv_test.go:646]: @digivava when you have a moment, can you give this PR another pass please?

thoughts on this test case?

**Review by digivava — APPROVED** [review]
This seems to work as expected now! Thank you!
